homeassistant/components/keyboard_remote/__init__.py

```diff
@@ -48,165 +60,210 @@
 )
 
 
-def setup(hass, config):
+async def async_setup(hass, config):
     """Set up the keyboard_remote."""
     config = config.get(DOMAIN)
 
-    keyboard_remote = KeyboardRemote(hass, config)
+    remote = KeyboardRemote(hass, config)
+    await remote.setup()
 
-    def _start_keyboard_remote(_event):
-        keyboard_remote.run()
+    return True
 
-    def _stop_keyboard_remote(_event):
-        keyboard_remote.stop()
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, _start_keyboard_remote)
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, _stop_keyboard_remote)
+class KeyboardRemote:
+    """Manage device connection/disconnection using inotify to asynchronously monitor."""
 
-    return True
+    def __init__(self, hass, config):
+        """Create handlers and setup dictionaries to keep track of them."""
+        self.hass = hass
+        self.handlers_by_name = {}
+        self.handlers_by_descriptor = {}
+        self.active_handlers_by_descriptor = {}
 
+        for dev_block in config:
+            handler = self.DeviceHandler(hass, dev_block)
+            descriptor = dev_block.get(DEVICE_DESCRIPTOR)
+            if descriptor is not None:
+                self.handlers_by_descriptor[descriptor] = handler
+            else:
+                name = dev_block.get(DEVICE_NAME)
+                self.handlers_by_name[name] = handler
+
+    async def setup(self):
+        """Start inotify watching for events, start event monitoring for those already connected, and start monitoring for device connection/disconnection."""
+
+        # start watching
+        self.watcher = aionotify.Watcher()
+        self.watcher.watch(
+            alias="devinput",
+            path=DEVINPUT,
+            flags=aionotify.Flags.CREATE
+            | aionotify.Flags.ATTRIB
+            | aionotify.Flags.DELETE,
+        )
+        await self.watcher.setup(self.hass.loop)
 
-class KeyboardRemoteThread(threading.Thread):
-    """This interfaces with the inputdevice using evdev."""
+        # add initial devices (do this AFTER starting watcher in order to avoid race conditions leading to missing device connections)
```

Comment:
Please break this comment on two lines for easier reading.

Reply:
done